File: api/typescript/src/core/lib/services/service_context.ts
import { err, ok, Result } from 'neverthrow';
import { newExecCommandArgs, newPauseServiceArgs, newUnpauseServiceArgs } from '../constructor_calls';
import type { ExecCommandArgs } from '../../kurtosis_core_rpc_api_bindings/api_container_service_pb';
import type { PortSpec } from './port_spec';
import type { ServiceName, ServiceUUID } from './service';
import { GenericApiContainerClient } from '../enclaves/generic_api_container_client';
import {PauseServiceArgs, UnpauseServiceArgs} from "../../kurtosis_core_rpc_api_bindings/api_container_service_pb";

// Docs available at https://docs.kurtosis.com/sdk/#servicecontext
export class ServiceContext {
    constructor(
        private readonly client: GenericApiContainerClient,
        private readonly serviceName: ServiceName,
        private readonly serviceUuid: ServiceUUID,
        private readonly privateIpAddress: string,
        private readonly privatePorts: Map<string, PortSpec>,
        private readonly publicIpAddress: string,
        private readonly publicPorts: Map<string, PortSpec>,
    ) {}

    // Docs available at https://docs.kurtosis.com/sdk/#getservicename---servicename
    public getServiceName(): ServiceName {
        return this.serviceName;
    }

    // Docs available at https://docs.kurtosis.com/sdk/#getserviceuuid---serviceuuid
    public getServiceUUID(): ServiceUUID {
        return this.serviceUuid;
    }

    // Docs available at https://docs.kurtosis.com/sdk/#getprivateipaddress---string
    public getPrivateIPAddress(): string {
        return this.privateIpAddress
    }

    // Docs available at https://docs.kurtosis.com/sdk/#getprivateports---mapportid-portspec
    public getPrivatePorts(): Map<string, PortSpec> {
        return this.privatePorts
    }

    // Docs available at https://docs.kurtosis.com/sdk/#getmaybepublicipaddress---string
    public getMaybePublicIPAddress(): string {
        return this.publicIpAddress
    }

    // Docs available at https://docs.kurtosis.com/sdk/#getpublicports---mapportid-portspec
    public getPublicPorts(): Map<string, PortSpec> {
        return this.publicPorts
    }

    // Docs available at https://docs.kurtosis.com/sdk/#execcommandliststring-command---int-exitcode-string-logs
    public async execCommand(command: string[]): Promise<Result<[number, string], Error>> {
        const execCommandArgs: ExecCommandArgs = newExecCommandArgs(this.serviceName, command);

        const execCommandResponseResult = await this.client.execCommand(execCommandArgs)
        if(execCommandResponseResult.isErr()){
            return err(execCommandResponseResult.error)
        }

        const execCommandResponse = execCommandResponseResult.value
        return ok([execCommandResponse.getExitCode(), execCommandResponse.getLogOutput()]);
    }
}

File: api/golang/core/lib/services/service_context.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package services

import (
	"context"

	"github.com/kurtosis-tech/kurtosis/api/golang/core/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis/api/golang/core/lib/binding_constructors"
	"github.com/kurtosis-tech/stacktrace"
)

type ServiceContext struct {
	client      kurtosis_core_rpc_api_bindings.ApiContainerServiceClient
	serviceName ServiceName
	serviceUuid ServiceUUID

	// Network location inside the enclave
	privateIpAddr string
	privatePorts  map[string]*PortSpec

	// Network location outside the enclave
	publicIpAddr string
	publicPorts  map[string]*PortSpec

	// Service labels
	labels map[string]string
}

func NewServiceContext(
	client kurtosis_core_rpc_api_bindings.ApiContainerServiceClient,
	serviceName ServiceName,
	serviceUuid ServiceUUID,
	privateIpAddr string,
	privatePorts map[string]*PortSpec,
	publicIpAddr string,
	publicPorts map[string]*PortSpec,
	labels map[string]string,
) *ServiceContext {
	return &ServiceContext{
		client:        client,
		serviceName:   serviceName,
		serviceUuid:   serviceUuid,
		privateIpAddr: privateIpAddr,
		privatePorts:  privatePorts,
		publicIpAddr:  publicIpAddr,
		publicPorts:   publicPorts,
		labels:        labels,
	}
}

func (service *ServiceContext) GetServiceName() ServiceName {
	return service.serviceName
}

func (service *ServiceContext) GetServiceUUID() ServiceUUID {
	return service.serviceUuid
}

func (service *ServiceContext) GetPrivateIPAddress() string {
	return service.privateIpAddr
}

func (service *ServiceContext) GetPrivatePorts() map[string]*PortSpec {
	return service.privatePorts
}

func (service *ServiceContext) GetMaybePublicIPAddress() string {
	return service.publicIpAddr
}

func (service *ServiceContext) GetPublicPorts() map[string]*PortSpec {
	return service.publicPorts
}

func (service *ServiceContext) GetLabels() map[string]string {
	return service.labels
}

func (service *ServiceContext) ExecCommand(command []string) (int32, string, error) {
	serviceName := service.serviceName
	args := binding_constructors.NewExecCommandArgs(string(serviceName), command)
	resp, err := service.client.ExecCommand(context.Background(), args)
	if err != nil {
		return 0, "", stacktrace.Propagate(
			err,
			"An error occurred executing command '%v' on service '%v'",
			command,
			serviceName)
	}
	return resp.ExitCode, resp.LogOutput, nil
}
